pkg/decision/evaluator/matchers/semver.go

```diff
@@ -0,0 +1,163 @@
+ * See the License for the specific language governing permissions and      *
+ * limitations under the License.                                           *
+ ***************************************************************************/
+
+// Package matchers //
+package matchers
+
+import (
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/optimizely/go-sdk/pkg/decision/reasons"
+	"github.com/optimizely/go-sdk/pkg/entities"
+
+	"github.com/pkg/errors"
+)
+
+// SemanticVersion defines the class
+type SemanticVersion struct {
+	Condition string // condition is always a string here
+}
+
+func (sv SemanticVersion) compareVersion(attribute string) (int, error) {
+
+	targetedVersionParts, err := sv.splitSemanticVersion(sv.Condition)
+	if err != nil {
+		return 0, err
+	}
+	versionParts, e := sv.splitSemanticVersion(attribute)
+	if e != nil {
+		return 0, e
+	}
+
+	// Up to the precision of targetedVersion, expect version to match exactly.
+	for idx := range targetedVersionParts {
+
+		switch {
+		case len(versionParts) <= idx:
+			return -1, nil
+		case !sv.isNumber(versionParts[idx]):
+			// Compare strings
+			if versionParts[idx] < targetedVersionParts[idx] {
+				return -1, nil
+			} else if versionParts[idx] > targetedVersionParts[idx] {
+				return 1, nil
+			}
+		case sv.isNumber(targetedVersionParts[idx]): // both targetedVersionParts and versionParts are digits
+			if sv.toInt(versionParts[idx]) < sv.toInt(targetedVersionParts[idx]) {
+				return -1, nil
+			} else if sv.toInt(versionParts[idx]) > sv.toInt(targetedVersionParts[idx]) {
+				return 1, nil
+			}
+		default:
+			return -1, nil
+		}
+	}
+
+	if sv.isPreRelease(attribute) && !sv.isPreRelease(sv.Condition) {
+		return -1, nil
+	}
+
+	return 0, nil
+}
+
+func (sv SemanticVersion) splitSemanticVersion(targetedVersion string) ([]string, error) {
+
+	if sv.hasWhiteSpace(targetedVersion) {
+		return []string{}, errors.New(string(reasons.AttributeFormatInvalid))
+	}
+
+	splitBy := ""
+	if sv.isBuild(targetedVersion) {
+		splitBy = sv.buildSeperator()
+	} else if sv.isPreRelease(targetedVersion) {
+		splitBy = sv.preReleaseSeperator()
+	}
+	targetParts := strings.Split(targetedVersion, splitBy)
+	if len(targetParts) == 0 {
+		return []string{}, errors.New(string(reasons.AttributeFormatInvalid))
+	}
+
+	targetPrefix := targetParts[0]
+	targetSuffix := targetParts[1:]
+
+	// Expect a version string of the form x.y.z
+	targetedVersionParts := strings.Split(targetPrefix, ".")
+
+	if len(targetedVersionParts) > 3 {
+		return []string{}, errors.New(string(reasons.AttributeFormatInvalid))
+	}
+
+	if len(targetedVersionParts) == 0 {
+		return []string{}, errors.New(string(reasons.AttributeFormatInvalid))
+	}
+
+	targetedVersionParts = append(targetedVersionParts, targetSuffix...)
+	return targetedVersionParts, nil
+}
+
+func (sv SemanticVersion) isNumber(str string) bool {
+	var digitCheck = regexp.MustCompile(`^[0-9]+$`)
```

Comment:
This can be a constant or at least defined globally.

Reply:
regexp.MustCompile cannot be constant, I moved it out of this function.